### ISSUE TITLE
Run integration tests against a working installation

### ISSUE THREAD

**Issue by csweichel** [gitpod-io/gitpod#6518]

**Comment by kylos101** [comment]
I learned how to manually run the integration tests and shared in #6599. 

For posterity, we ran the tests a few times last week against a self-hosted installation, the results are below. 

- [ ] Next steps, create issues to address the test failures and prioritize with the team. 
  * There is related work happening in #5985.

## Run 1
### Summary (16 failures):
9 tests failed due to being unable to parse  DB_PORT. (we'll need to [extend the test framework](https://github.com/gitpod-io/gitpod/blob/a9dca2a2bf7f1562ee90e53b038d63f9cbba824f/test/pkg/integration/apis.go#L432) as the DB port comes from a secret now)
2 tests failed due to OutOfcpu.
1 test failed due to minio.default.svc.cluster.local: no such host
4 tests failed due to cannot start workspace, this is a known issue impacting the saas, too.

## Run 2:
### Summary (16 failures)"
Same as **Run 1**, except instead of the two **OutOfCpu**, we got a couple builder_test failures:
`builder_test.go:84: image build failed: image build did not produce a workspace image`
`builder_test.go:194: image builder error: rpc error: code = DeadlineExceeded desc = context deadline exceeded`

```
    content-service_test.go:273: HTTP PUT request failed: "Put \"http://minio.default.svc.cluster.local:9000/gitpod-user-00000000-0000-0000-0000-000000000000/blobs/test-blob?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=GtPfiEfWkDL0A9Q2r9Vh%2F20211104%2Flocal%2Fs3%2Faws4_request&X-Amz-Date=20211104T184109Z&X-Amz-Expires=1800&X-Amz-SignedHeaders=host&X-Amz-Signature=25f312a1adf9bb98171eb430bd29205334e716c84d3c37584cd4c05676bc7bc9\": dial tcp: lookup minio.default.svc.cluster.local: no such host"
--- FAIL: TestUploadDownloadBlob (2.18s)
    --- FAIL: TestUploadDownloadBlob/UploadDownloadBlob (2.18s)
        --- FAIL: TestUploadDownloadBlob/UploadDownloadBlob/it_should_upload_and_download_blob (2.18s)
--
    content-service_test.go:234: cannot get content blob upload URL: "cannot access Gitpod server API: \"error parsing DB_PORT '' on pod server-54cf8cd79-xtfbx!\""
--- FAIL: TestUploadDownloadBlobViaServer (0.14s)
    --- FAIL: TestUploadDownloadBlobViaServer/UploadDownloadBlobViaServer (0.14s)
        --- FAIL: TestUploadDownloadBlobViaServer/UploadDownloadBlobViaServer/it_should_uploads_a_blob_via_server_→_content-server_and_downloads_it_afterwards (0.14s)
--
    db_test.go:32: error parsing DB_PORT '' on pod server-54cf8cd79-xtfbx!
--- FAIL: TestBuiltinUserExists (0.13s)
    --- FAIL: TestBuiltinUserExists/database (0.13s)
        --- FAIL: TestBuiltinUserExists/database/it_should_exists_a_builtin_user_workspace (0.13s)
--
    builder_test.go:84: image build failed: OutOfcpu: Pod Node didn't have enough resource: cpu, requested: 1000, used: 3819, capacity: 3920
--- FAIL: TestBaseImageBuild (2.34s)
    --- FAIL: TestBaseImageBuild/database (2.34s)
        --- FAIL: TestBaseImageBuild/database/it_should_build_a_base_image (2.34s)
--
    builder_test.go:194: image build failed: OutOfcpu: Pod Node didn't have enough resource: cpu, requested: 1000, used: 3819, capacity: 3920
--- FAIL: TestParallelBaseImageBuild (280.18s)
    --- FAIL: TestParallelBaseImageBuild/image-builder (280.18s)
        --- FAIL: TestParallelBaseImageBuild/image-builder/it_should_allow_parallel_build_of_images (280.18s)
--
    blockuser_test.go:39: cannot create user: "error parsing DB_PORT '' on pod server-54cf8cd79-xtfbx!"
--- FAIL: TestAdminBlockUser (0.14s)
    --- FAIL: TestAdminBlockUser/block_user (0.14s)
        --- FAIL: TestAdminBlockUser/block_user/it_should_block_new_created_user (0.14s)
--
    server_test.go:33: cannot get GitpodServer: "cannot access Gitpod server API: \"error parsing DB_PORT '' on pod server-54cf8cd79-xtfbx!\""
--- FAIL: TestServerAccess (0.14s)
    --- FAIL: TestServerAccess/GetLoggedInUser (0.14s)
        --- FAIL: TestServerAccess/GetLoggedInUser/it_can_get_a_not_built-in_logged_user (0.14s)
--
    server_test.go:62: cannot get GitpodServer: "cannot access Gitpod server API: \"error parsing DB_PORT '' on pod server-54cf8cd79-xtfbx!\""
--- FAIL: TestStartWorkspace (0.14s)
    --- FAIL: TestStartWorkspace/CreateWorkspace (0.14s)
        --- FAIL: TestStartWorkspace/CreateWorkspace/it_can_run_workspace_tasks (0.14s)
--
    content_test.go:41: cannot start workspace: "cannot wait for workspace: \"context deadline exceeded\""
--- FAIL: TestBackup (300.01s)
    --- FAIL: TestBackup/backup (300.01s)
        --- FAIL: TestBackup/backup/it_should_start_a_workspace,_create_a_file_and_successfully_create_a_backup (300.01s)
--
    contexts_test.go:120: cannot start server: "cannot access Gitpod server API: \"error parsing DB_PORT '' on pod server-54cf8cd79-xtfbx!\""
--- FAIL: TestGitHubContexts (0.15s)
    --- FAIL: TestGitHubContexts/context (0.15s)
        --- FAIL: TestGitHubContexts/context/should_run_context_tests (0.00s)
            --- FAIL: TestGitHubContexts/context/should_run_context_tests/github.com/gitpod-io/gitpod-test-repo/tree/integration-test-1 (0.14s)
            --- FAIL: TestGitHubContexts/context/should_run_context_tests/github.com/gitpod-io/gitpod-test-repo/tree/integration-test-context-tag (0.14s)
            --- FAIL: TestGitHubContexts/context/should_run_context_tests/github.com/gitpod-io/gitpod-test-repo/issues/88 (0.14s)
            --- FAIL: TestGitHubContexts/context/should_run_context_tests/github.com/gitpod-io/gitpod (0.15s)
--
    contexts_test.go:120: cannot start server: "cannot access Gitpod server API: \"error parsing DB_PORT '' on pod server-54cf8cd79-xtfbx!\""
--- FAIL: TestGitLabContexts (0.41s)
    --- FAIL: TestGitLabContexts/context (0.41s)
        --- FAIL: TestGitLabContexts/context/should_run_context_tests (0.00s)
            --- FAIL: TestGitLabContexts/context/should_run_context_tests/gitlab.com/AlexTugarev/gp-test/issues/1 (0.14s)
            --- FAIL: TestGitLabContexts/context/should_run_context_tests/gitlab.com/AlexTugarev/gp-test (0.14s)
            --- FAIL: TestGitLabContexts/context/should_run_context_tests/gitlab.com/AlexTugarev/gp-test/tree/wip (0.21s)
            --- FAIL: TestGitLabContexts/context/should_run_context_tests/gitlab.com/AlexTugarev/gp-test/merge_requests/2 (0.41s)
--
    docker_test.go:33: cannot start workspace: "workspace instance c6f74404-9356-415e-9410-26d41f260a7b has stopped"
--- FAIL: TestRunDocker (5.86s)
    --- FAIL: TestRunDocker/docker (5.86s)
        --- FAIL: TestRunDocker/docker/it_should_start_a_container (5.86s)
--
    example_test.go:33: cannot start server: "cannot access Gitpod server API: \"error parsing DB_PORT '' on pod server-54cf8cd79-xtfbx!\""
--- FAIL: TestWorkspaceInstrumentation (0.14s)
    --- FAIL: TestWorkspaceInstrumentation/instrumentation (0.14s)
        --- FAIL: TestWorkspaceInstrumentation/instrumentation/it_can_instrument_a_workspace (0.14s)
--
    example_test.go:76: cannot start workspace: "workspace instance d80249f7-b85b-4b48-82d2-07f505280ed3 has stopped"
--- FAIL: TestLaunchWorkspaceDirectly (5.36s)
    --- FAIL: TestLaunchWorkspaceDirectly/workspace (5.36s)
        --- FAIL: TestLaunchWorkspaceDirectly/workspace/it_can_run_workspace_tasks (5.36s)
--
    ghost_test.go:41: cannot start workspace: "workspace instance db522dbe-b161-4963-b4c9-eda51489aacf has stopped"
--- FAIL: TestGhostWorkspace (14.12s)
    --- FAIL: TestGhostWorkspace/ghost (14.12s)
        --- FAIL: TestGhostWorkspace/ghost/it_can_start_a_ghost_workspace (14.12s)
--
    git_test.go:118: cannot start server: "cannot access Gitpod server API: \"error parsing DB_PORT '' on pod server-54cf8cd79-xtfbx!\""
=== RUN   TestGitActions/GitActions/it_can_run_git_actions/github.com/gitpod-io/gitpod-test-repo/tree/integration-test/commit-and-push#01
--- FAIL: TestGitActions (0.13s)
    --- FAIL: TestGitActions/GitActions (0.13s)
        --- FAIL: TestGitActions/GitActions/it_can_run_git_actions (0.13s)
            --- FAIL: TestGitActions/GitActions/it_can_run_git_actions/github.com/gitpod-io/gitpod-test-repo/tree/integration-test/commit-and-push (0.13s)
```

**Comment by ghuntley** [comment]
Moin, @kylos101. Here's some insights I've learned which might be useful to you!

Integration tests of the context parsers are controlled via setting the tokens:
```
GITPOD_TEST_TOKEN_GITHUB at https://github.com/gitpod-io/gitpod/blob/335215f565e2912d5d84db3f1502f30ed53d24fd/components/server/src/github/github-context-parser.spec.ts#L30
GITPOD_TEST_TOKEN_GITLAB at https://github.com/gitpod-io/gitpod/blob/335215f565e2912d5d84db3f1502f30ed53d24fd/components/server/src/gitlab/gitlab-context-parser.spec.ts#L24
GITPOD_TEST_TOKEN_BITBUCKET at https://github.com/gitpod-io/gitpod/blob/335215f565e2912d5d84db3f1502f30ed53d24fd/components/server/src/bitbucket/bitbucket-context-parser.spec.ts#L21
```

**Comment by kylos101** [comment]
> Moin, @kylos101. Here's some insights I've learned which might be useful to you!
> 
> Integration tests of the context parsers are controlled via setting the tokens:
> 
> ```
> GITPOD_TEST_TOKEN_GITHUB at https://github.com/gitpod-io/gitpod/blob/335215f565e2912d5d84db3f1502f30ed53d24fd/components/server/src/github/github-context-parser.spec.ts#L30
> GITPOD_TEST_TOKEN_GITLAB at https://github.com/gitpod-io/gitpod/blob/335215f565e2912d5d84db3f1502f30ed53d24fd/components/server/src/gitlab/gitlab-context-parser.spec.ts#L24
> GITPOD_TEST_TOKEN_BITBUCKET at https://github.com/gitpod-io/gitpod/blob/335215f565e2912d5d84db3f1502f30ed53d24fd/components/server/src/bitbucket/bitbucket-context-parser.spec.ts#L21
> ```

Thank you, @ghuntley ! 

It looks like there are 7 Typescript spec files that have integration tests:
* 6 tests for Github (1) / Gitlab (1) / Bitbucket (4) using the tokens you mentioned. 
* 1 test for the DB, CodeSyncResourceDBSpec, which has 10s timeout

Some follow-up questions (I'll ask these in **help-me**, too):
1. Are there any other integration tests I have missed? I could not find others.
2. Are these tests running in Werft? If no, should they? Depending on your answer, we may want to create a separate issue and route to Team Meta.
   * I would not expect the 6 ContextParser tests to be running in Werft. 
   * The `@suite` annotation on this will skip the test if it cannot find the related ENV VAR.
   * The injection of these tokens as ENV VARs appears to be commented out in [debug](https://github.com/gitpod-io/gitpod/blob/3729245a7f37d11c36defd460d697f5cd1892fb2/.werft/debug.yaml#L121-L135) and [build](https://github.com/gitpod-io/gitpod/blob/a17cc00b3a4d3622eb57c359a027db81f1123208/.werft/build.yaml#L121-L135) configurations. e.g. `skipIfEnvVarNotSet("GITPOD_TEST_TOKEN_BITBUCKET")`.
3. How do you run these integration tests manually now? I assume you set the ENV VAR in your workspace, and then run `npm test` or `yarn test`?

Also, as a heads up, we can run a variety of tests for [Github](https://github.com/gitpod-io/gitpod/blob/main/test/tests/workspace/contexts_test.go#L29), and [Gitlab](https://github.com/gitpod-io/gitpod/blob/main/test/tests/workspace/contexts_test.go#L61), but not Bitbucket, using Go tests.

**Comment by ghuntley** [comment]
As the integration tests are hot and not mocked then there will be API quota tokens. Thus suggesting running this twice a day on cron against main?

**Comment by kylos101** [comment]
Hi @csweichel , @MrSimonEmms ,

It takes about 10 minutes to [run the integration tests](https://github.com/gitpod-io/gitpod/tree/main/test#manually-against-a-kubernetes-cluster) against a self hosted installation. 

The below failures are present for self-hosted. I thought resolving the DB_PORT issue would bring the # of failures down, but, I was wrong. 

I plan to dig into these some more tomorrow, before starting work on using the installer with core-dev, and wanted to give you both a heads up. In other words, the # of failures with core-dev for staging-main is 4-5. If we start using the installer with core-dev now, that # of failures will be higher (about 15-16).

```
    content-service_test.go:275: HTTP PUT request failed: "...  dial tcp: lookup minio.default.svc.cluster.local: no such host"
--- FAIL: TestUploadDownloadBlob (2.18s) new	

    content-service_test.go:275: HTTP PUT request failed: ...  dial tcp: lookup minio.default.svc.cluster.local: no such host"
--- FAIL: TestUploadDownloadBlobViaServer (4.42s)  new

    builder_test.go:84: image build failed: image build did not produce a workspace image
--- FAIL: TestBaseImageBuild (173.81s)  new

    blockuser_test.go:71: cannot perform AdminBlockUser: "jsonrpc2: code 555 message: enterprise license required"
--- FAIL: TestAdminBlockUser (17.53s)  new

    server_test.go:96: cannot get workspace: "cannot wait for workspace: \"context deadline exceeded\""
--- FAIL: TestStartWorkspace (300.42s)  new

    content_test.go:41: cannot start workspace: "workspace instance 181af11b-34c2-478e-8f1f-b196e0181fce has stopped"
--- FAIL: TestBackup (5.25s)  new

    contexts_test.go:132: timed out attempting to connect agent: dial-http tcp localhost:43853: read tcp 127.0.0.1:52724->127.0.0.1:43853: read: connection reset by peer
--- FAIL: TestGitHubContexts (131.32s)  new
```

**Comment by kylos101** [comment]
> Hi @csweichel , @MrSimonEmms ,
> 
> It takes about 10 minutes to [run the integration tests](https://github.com/gitpod-io/gitpod/tree/main/test#manually-against-a-kubernetes-cluster) against a self hosted installation.
> 
> The below failures are present for self-hosted. I thought resolving the DB_PORT issue would bring the # of failures down, but, I was wrong.
> 
> I plan to dig into these some more tomorrow, before starting work on using the installer with core-dev, and wanted to give you both a heads up. In other words, the # of failures with core-dev for staging-main is 4-5. If we start using the installer with core-dev now, that # of failures will be higher (about 15-16).
> 
> ```
>     content-service_test.go:275: HTTP PUT request failed: "...  dial tcp: lookup minio.default.svc.cluster.local: no such host"
> --- FAIL: TestUploadDownloadBlob (2.18s) new	
> 
>     content-service_test.go:275: HTTP PUT request failed: ...  dial tcp: lookup minio.default.svc.cluster.local: no such host"
> --- FAIL: TestUploadDownloadBlobViaServer (4.42s)  new
> 
>     builder_test.go:84: image build failed: image build did not produce a workspace image
> --- FAIL: TestBaseImageBuild (173.81s)  new
> 
>     blockuser_test.go:71: cannot perform AdminBlockUser: "jsonrpc2: code 555 message: enterprise license required"
> --- FAIL: TestAdminBlockUser (17.53s)  new
> 
>     server_test.go:96: cannot get workspace: "cannot wait for workspace: \"context deadline exceeded\""
> --- FAIL: TestStartWorkspace (300.42s)  new
> 
>     content_test.go:41: cannot start workspace: "workspace instance 181af11b-34c2-478e-8f1f-b196e0181fce has stopped"
> --- FAIL: TestBackup (5.25s)  new
> 
>     contexts_test.go:132: timed out attempting to connect agent: dial-http tcp localhost:43853: read tcp 127.0.0.1:52724->127.0.0.1:43853: read: connection reset by peer
> --- FAIL: TestGitHubContexts (131.32s)  new
> ```

@csweichel  @MrSimonEmms I was surprised at some of these failures, like TestAdminBlockUser, which I haven't seen before with self hosted.

**Comment by kylos101** [comment]
This is also related to #5317
